ci/enumerate_build_matrix.sh

```diff
@@ -12,7 +12,28 @@ set -x # be noisy + log everything that is happening in the script
 
 function runmake()
 {
-	make -s clean all $*
+  make -s all $*
+}
+
+export BUILD_JOBS_DIRECTORY="$PWD/build/jobs"
+
+function runBuildJob()
+{
+  build_directory="${BUILD_JOBS_DIRECTORY}/$2"
+  make_directory="$PWD/$1"
+  echo
+  echo '-----------------------------------------------------------------------'
+  echo "Running ${@:3} in $1, build directory ${build_directory}"
+  mkdir -p "${build_directory}"
+  cd "${BUILD_JOBS_DIRECTORY}"
+  eval ${@:3} -C "${make_directory}" BUILD_PATH_BASE="${build_directory}"
+  # I didn't want to deal with another build type, so we are specifically dealing
+  # with newhal build here
+  if echo "${@:3}" | grep --quiet "newhal"; then
```

Comment:
Can you describe why so many changes were necessary for `hal/newhal` and is the above just moving the `newhal` build to the top of the process before the jobs run in parallel?  Are we still building the newhal platform?

FYI, this test for `has no sections` as a way to gauge pass/fail seems to have broken down... looking back at a previous build:

```
 make -s clean all PLATFORM=newhal COMPILE_LTO=n
runing matrix PLATFORM= MODULAR_PLATFORM= PLATFORM_BOOTLOADER=
-----------------------------------------------------------------------
In file included from ../wiring/inc/spark_wiring_cloud.h:32:0,
                 from ../wiring/inc/spark_wiring.h:51,
                 from ./inc/application.h:40,
                 from src/application.cpp:21:
../wiring/inc/spark_wiring_async.h:32:2: error: #error "std::atomic is not always lock-free for required types"
 #error "std::atomic is not always lock-free for required types"
  ^
make[1]: *** [../build/target/user/platform-60000/src/src/application.o] Error 1
../build/module.mk:277: recipe for target '../build/target/user/platform-60000/src/src/application.o' failed
../build/recurse.mk:11: recipe for target 'user' failed
make: *** [user] Error 2
++ echo 2
++ grep 'has no sections'
+ HAS_NO_SECTIONS=
+ [[ ! -z HAS_NO_SECTIONS ]]
+ testcase
+ '[' 0 -eq 0 ']'
 ✓ SUCCESS 
```

Clearly it failed to build, but we passed it.

Reply:
It's become out of date with the reset of the HALs (no implementation headers etc.) and we've simply never actually checked that it builds.

I agree that this whole `has no sections` thing seems weird and probably never worked. For some reason we are grepping the result code for that string, instead of the stdout/stderr.

Reply:
And yes, we are building `newhal` normally like all the other things. You can see that it's added to the list of jobs [here](https://github.com/particle-iot/device-os/pull/1712/files/a7215e55fe1d29e858e9bae3196fbcd42d2ef26c#diff-a3c50403e095e3bff90f2e7ed13f36daR85) which are then later run in parallel [here](https://github.com/particle-iot/device-os/pull/1712/files/a7215e55fe1d29e858e9bae3196fbcd42d2ef26c#diff-a3c50403e095e3bff90f2e7ed13f36daR166)

Reply:
`has no sections` was an additional check for error, since the result code of 0 with `has no sections` is not a passing case: https://github.com/particle-iot/device-os/commit/b0dde1cb0d0faa54c1f4dfef1b1c22a62b6daead#diff-a3c50403e095e3bff90f2e7ed13f36da

Looks like Mat chopped out the pass/fall portion later during the parallel builds (not sure why).

Reply:
^ Unless I'm misinterpreting how `testcase` works... 

From the build logs... I'm not sure newhal is compiling (notice no output under it):
```
-----------------------------------------------------------------------
Running runmake PLATFORM="newhal" COMPILE_LTO="n" in main, build directory /firmware/build/jobs/0
-----------------------------------------------------------------------
Running runmake DEBUG_BUILD="n" PLATFORM="xenon" COMPILE_LTO="n" SPARK_CLOUD="y" in main, build directory /firmware/build/jobs/1
Creating /firmware/build/jobs/1/user-part/platform-14-m/module_user_memory.ld ...
Creating /firmware/build/jobs/1/user-part/platform-14-m/module_user_memory.ld ...
   text	   data	    bss	    dec	    hex	filename
  11012	    152	   1060	  12224	   2fc0	/firmware/build/jobs/1/user-part/platform-14-m/user-part.elf
 ✓ SUCCESS 
```

Reply:
The builds are run in parallel and the stdout/stderr is interleaved. It's there somewhere: https://travis-ci.org/particle-iot/device-os/jobs/503331630#L1367

Reply:
I've fixed the check in this commit: https://github.com/particle-iot/device-os/pull/1712/commits/a7c21d5f55af7c32f127b11aa581727df014670a

Reply:
Still seems to be failing (and passing anyway), but we can fix this one later since there are more important things to test!!  I'll create a story for this one.
```
-----------------------------------------------------------------------
Running runmake PLATFORM="newhal" COMPILE_LTO="n" in main, build directory /firmware/build/jobs/0
arm-none-eabi-objcopy: error: the input file '/firmware/build/jobs/0/main/platform-60000/main.elf' has no sections
../build/module.mk:216: recipe for target '/firmware/build/jobs/0/main/platform-60000/main.bin' failed
make: *** [/firmware/build/jobs/0/main/platform-60000/main.bin] Error 1
 ✓ SUCCESS 
```